contents/tutorials/bootstrap-feature-flags-react.md

```diff
@@ -0,0 +1,317 @@
+
+## Set up the React app for server-side rendering
+
+To bootstrap our flags, we fetch the feature flag data on the backend and pass it to the frontend before it loads. This requires server-side rendering our React app. To do this, we set up:
+
+1. The bundler and build the React app.
+2. An Express server to get feature flag data from PostHog and serve the React app.
+
+To start, install the package to compile, transform assets, and hot reload the server on file change.
+
+```bash
+npm install --save-dev @babel/register babel-plugin-transform-assets
+npm install -g nodemon
+```
+
+After this, we make one more change to `index.js` to change `ReactDOM.createRoot` to `ReactDOM.hydrateRoot`.
+
+```js
+// src/index.js
+//...
+const root = ReactDOM.hydrateRoot(document.getElementById('root'));
+root.render(
+  <React.StrictMode>
+    <App />
+  </React.StrictMode>
+);
+```
+
+Finally, build the client so we can serve it from the Express server.
+
+```bash
+npm run build
+```
+
+## Set up our server-rendering Express app
+
+To start, create another folder named `server` in the client folder. Inside this folder, create an `index.js` file. In this file, we:
+
+1. Set up Babel.
+2. Import the packages for React, Express, and the file system.
+3. Set up an Express get handler for all routes that:
+    1. Creates a React app from our built client code
+    2. Reads the index file from the built client code
+    3. Replaces the div root with our server-created React app
+
+```js
+require('@babel/register')({
+  presets: [
+    '@babel/preset-env',
+    ['@babel/preset-react', { runtime: 'automatic' }],
+  ],
+  plugins: [
+    [
+      'transform-assets',
+      {
+        extensions: ['css', 'svg'],
+        name: 'static/media/[name].6ce24c58023cc2f8fd88fe9d219db6c6.[ext]',
+      },
+    ],
+  ],
+});
+
+const React = require('react');
+const ReactDOMServer = require('react-dom/server');
+const App = require('../src/App').default;
+const express = require('express');
+const path = require('path');
+const fs = require('fs');
+
+const app = express();
+
+app.get('/*', (req, res, next) => {
+  if (req.url !== '/') {
+    return next();
+  }
+  const reactApp = ReactDOMServer.renderToString(React.createElement(App));
+
+  const indexFile = path.resolve('build/index.html');
+  fs.readFile(indexFile, 'utf8', (err, data) => {
+    if (err) {
+      return res.status(500).send(err);
+    }
+
+    return res.send(
+      data.replace('<div id="root">', `<div id="root">${reactApp}`)
+    );
+  });
+});
+
+app.use(express.static(path.resolve(__dirname, '../build')));
+
+app.listen(8080, () =>
+  console.log('Express server is running')
+);
+```
+
+Once done, you can run the server with `nodemon server` and see your site in action (it looks the same).
+
+```bash
+nodemon server
+```
+
+## Handle feature flags on the backend
+
+To bootstrap the flags, first, we need the flag data from PostHog. To do this, we install `posthog-node`.
```

Comment:
One slight problem here:

If the backend is going to our servers to fetch flags anyway, then bootstrapping isn't adding much since you're still waiting for flags to load.

In this case, you could easily just not show anything on your client-side react app until all flags are loaded (by leveraging posthog.onFeatureFlags() ) - and it would be a simpler alternative to server-side rendering + getting flags from a backend library.

What makes bootstrapping specially useful is when you don't have this delay - which you get by local evaluation in posthog-node!

So, I'd suggest adding this bit here, and then: 

(1) initialise posthog-node with a person API key, so local eval is enabled
(2) pass in person properties / group properties to `getAllFlags()` so all flags that are required can be computed without going to posthog's servers.